src/sinks/util/service/health.rs

```diff
@@ -0,0 +1,324 @@
+                        .unwrap_or(RETRY_MAX_DURATION_SECONDS_DEFAULT),
+                )),
+        }
+    }
+}
+
+pub trait HealthLogic: Clone + Send + Sync + 'static {
+    type Error: Send + Sync + 'static;
+    type Response;
+
+    /// Returns health of the endpoint based on the response/error.
+    /// None if there is not enough information to determine it.
+    fn is_healthy(&self, response: &Result<Self::Response, Self::Error>) -> Option<bool>;
+}
+
+enum CircuitState {
+    /// Service is unhealthy hence it's not passing requests downstream.
+    /// Contains timeout.
+    Open(BoxFuture<'static, ()>),
+
+    /// Service will pass one request to test its health.
+    HalfOpen {
+        permit: Option<Trigger>,
+        done: Tripwire,
+    },
+
+    /// Service is healthy and passing requests downstream.
+    Closed(OpenToken<fn(usize)>),
+}
+
+/// A service which monitors the health of a service.
+/// Behaves like a circuit breaker.
+pub struct HealthService<S, L> {
+    inner: S,
+    logic: L,
+    counters: Arc<HealthCounters>,
+    snapshot: HealthSnapshot,
+    backoff: ExponentialBackoff,
+    state: CircuitState,
+    open: OpenGauge,
+    endpoint: String,
+}
+
+impl<S, L, Req> Service<Req> for HealthService<S, L>
+where
+    L: HealthLogic<Response = S::Response, Error = S::Error>,
+    S: Service<Req>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = HealthFuture<S::Future, L>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        loop {
+            self.state = match self.state {
+                CircuitState::Open(ref mut timer) => {
+                    ready!(timer.as_mut().poll(cx));
+
+                    debug!(message="Endpoint is on probation.", endpoint = %&self.endpoint);
+
+                    // Using Tripwire will let us be notified when the request is done.
+                    // This can't be done through counters since a requests can end without changing them.
+                    let (permit, done) = Tripwire::new();
+
+                    CircuitState::HalfOpen {
+                        permit: Some(permit),
+                        done,
+                    }
+                }
+                CircuitState::HalfOpen {
+                    permit: Some(_), ..
+                } => {
+                    // Pass one request to test health.
+                    return self.inner.poll_ready(cx).map_err(Into::into);
+                }
+                CircuitState::HalfOpen {
+                    permit: None,
+                    ref mut done,
+                } => {
+                    let done = Pin::new(done);
+                    ready!(done.poll(cx));
+
+                    if self.counters.healthy(self.snapshot).is_ok() {
+                        // A healthy response was observed
+                        info!(message="Endpoint is healthy.", endpoint = %&self.endpoint);
```

Comment:
```suggestion
                        info!("Endpoint is healthy.", endpoint = %&self.endpoint);
```

Reply:
Seems like this could be really noisy, `debug` seems a better level @ktff

Reply:
It seems like a useful information for users but true it could get noisy so maybe a `internal_log_rate_secs` of 5 seconds? That could also be added to `warn!(message = "Endpoint is unhealthy.", endpoint = %&self.endpoint);`

Reply:
Although I'm not sure if that can mess up the ordering of those messages which is important.

Reply:
@spencergilbert here is the log output for elasticsearch with this code when errors start occuring

```
2022-08-22T12:19:15.021985Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=10}:http: vector::internal_events::http_client: HTTP error. error=connection closed before message completed error_type="request_failed" stage="processing"
2022-08-22T12:19:15.022108Z  WARN sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=10}: vector::sinks::util::retries: Retrying after error. error=Failed to make HTTP(S) request: connection closed before message completed
2022-08-22T12:19:15.034182Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=11}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:15.034279Z  WARN sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=11}: vector::sinks::util::retries: Retrying after error. error=Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 111)
2022-08-22T12:19:15.043026Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=12}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:15.043093Z  WARN sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=12}: vector::sinks::util::retries: Retrying after error. error=Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 111)
2022-08-22T12:19:15.056103Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=13}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:15.056203Z  WARN sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=13}: vector::sinks::util::retries: Retrying after error. error=Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 111)
2022-08-22T12:19:15.069006Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=14}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:15.069095Z  WARN sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:request{request_id=14}: vector::sinks::util::retries: Retrying after error. error=Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 111)
2022-08-22T12:19:15.069579Z  WARN sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}: vector::sinks::util::service::health: Endpoint is unhealthy. endpoint=http://localhost:9201
2022-08-22T12:19:16.071767Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:18.073956Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:22.075408Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:30.077493Z ERROR sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 111) error_type="request_failed" stage="processing"
2022-08-22T12:19:48.401471Z  INFO sink{component_kind="sink" component_id=es component_type=elasticsearch component_name=es}: vector::sinks::util::service::health: Endpoint is healthy. endpoint=http://localhost:9201
```

It seems like more than adding to the noise logging `Endpoint healthy` as info helps to communicate that the errors were transient and that everything is fine now.

Reply:
Yeah, I think WARN is the right level for the logs in this code, and it's probably worth examining the ERRORs in that output, as part of a separate issue/PR, in terms of not emitting them so frequently, etc.